build-system/compile/internal-version.js

```diff
@@ -26,40 +26,77 @@ const argv = minimist(process.argv.slice(2), {
+  // - Find the commit time of <C> (<C>.time). Note that commit time might be
+  //   different from author time! e.g., commit time might be the time that a PR
+  //   was merged into `master`, or a commit was cherry-picked onto the branch;
+  //   author time is when the original author of the commit ran the
+  //   "git commit" command.
+  // - The version number is <C>.time.format("YYmmDDHHMM", "UTC") + <X> (the
+  //   pseudo-code assumes standard `.strftime` formatting, and <X> gets
+  //   zero-padded until it is three digits long).
+  //   - The maximum number of cherry-picks in a single branch is 999. More than
+  //     that will fail to build. This should never happen.
+  //
+  // Examples:
+  // 1. The version number of a release built from the HEAD commit on `master`,
+  //    where that HEAD commit was committed on April 25, 2020 2:31 PM EDT would
+  //    be `2004251831000`.
+  //    - EDT is UTC-4, so the hour value changes from EDT's 14 to UTC's 18.
+  //    - The last three digits are 000 as this commit is on `master`.
+  //
+  // 2. The version number of a release built from a local working branch (e.g.,
+  //    on a developer's workstation) that was split off from a `master` commit
+  //    from May 6, 2021 10:40 AM PDT and has multiple commits that exist only
+  //    on local working branch would be `2105061840000`.
+  //    - PDT is UTC-7, so the hour value changes from PDT's 10 to UTC's 17.
+  //    - The last three digits are 000 as this commit is on a branch that was
+  //      split from `master`, and does not have any cherry-picked commits.
+  //
+  // 3. For a release built from a local working branch that was split off from
+  //    a `master` commit from November 9, 2021 11:48 PM PST, and then:
+  //    - had one commit that was cherry-picked from `master`,
+  //    - followed by two commits that were created directly on the branch, the
+  //      last of which was commited on November 23, 2021 6:38 PM PST,
+  //    - followed by twelve commits that were cherry-picked from `master`,
+  //    then its version number would be `2111240238012`.
+  //    - The latest twelve commits are cherry-picks from `master`, and the one
+  //      before them is not, so our last three digits are set to 012.
+  //    - PST is UTC-8, so the hour value changes from PST's 18 to UTC's 2, and
+  //      one day is added.
+  let numberOfCherryPicks = 0;
+  const commitCherriesInfo = gitCherryMaster().reverse();
+  for (const {isCherryPick} of commitCherriesInfo) {
+    if (!isCherryPick) {
+      break;
     }
-    return `${lastCommitFormattedTime}${lastDigit}`;
+    numberOfCherryPicks++;
+  }
+  if (numberOfCherryPicks > 999) {
+    throw new Error(
+      `This branch has ${numberOfCherryPicks} cherry-picks, which is more than 999, the maximum allowed number of cherry-picks!`
+    );
   }
+
+  const lastCommitFormattedTime = gitCommitFormattedTime(
+    `HEAD~${numberOfCherryPicks}`
```

Comment:
Sanity check: Does this count the number of commits on the CP branch? Or does it count just  the number of cherry-picked commits?

In other words, I think this should count both `+` and `-` prefixed commits in order to get to the branch point, right?

Reply:
I was worried about the case where somebody is restarting an old PR and does a merge commit on their working branch - suddenly there could be thousands of commits in between because of the merge

The method in the git helper module does return all of them, and can be used somewhere else to verify that a release only contains commits directly,or cherry-picked, from `master`

Reply:
My concern is whether we're using the timestamp of the correct commit here. 

Suppose there are 3 cherry-picked commits and 4 new commits in a CP branch. Shouldn't this function use the timestamp of `HEAD~7` and not `HEAD~3`? Or could it be that I'm missing something?

Reply:
Just ran some tests on my local branch, and this is indeed the wrong timestamp. (In the example above, it should be `HEAD~7`, not `HEAD~3`.)

Reply:
I designed it specifically to use the last "new" commit (aka not a cherry-picked commit) instead of using the branch point.

Imagine a situation where some creates a branch from release `1234567890000` and adds 1 new commit to it. Do we want their version number to *also* be `123456789000`? There are no cherry picks, we want to use the time of the last new commit in the branch for that :)

Reply:
(also this is exactly example 3 in the doc)

Reply:
Thanks for explaining. For some reason, that last point was lost to me. Also, when I tested your branch locally, I had to include your commits, which show up as "new" commits and changed the calculation. 🤦